ARCs/arc-0056.md

```diff
@@ -0,0 +1,381 @@
+    type: ABIType;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+    /** Optional, user-friendly description for the return value */
+    desc?: string;
+  };
+  /** an action is a combination of call/create and an OnComplete */
+  actions: {
+    /** OnCompletes this method allows when appID === 0 */
+    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
+    /** OnCompletes this method allows when appID !== 0 */
+    call: (
+      | "NoOp"
+      | "OptIn"
+      | "CloseOut"
+      | "ClearState"
+      | "UpdateApplication"
+      | "DeleteApplication"
+    )[];
+  };
+  /** If this method does not write anything to the ledger (ARC-22) */
+  readonly?: boolean;
+  /** ARC-28 events that MAY be emitted by this method */
+  events?: Array<Event>;
+  /** Information that clients can use when calling the method */
+  recommendations?: {
+    /** The number of inner transactions the caller should cover the fees for */
+    innerTransactionCount?: number;
+    /** Recommended box references to include */
+    boxes?: {
+      /** The app ID for the box */
+      app?: number;
+      /** The base64 encoded box key */
+      key: string;
+      /** The number of bytes being read from the box */
+      readBytes: number;
+      /** The number of bytes being written to the box */
+      writeBytes: number;
+    };
+    /** Recommended foreign accounts */
+    accounts?: string[];
+    /** Recommended foreign apps */
+    apps?: number[];
+    /** Recommended foreign assets */
+    assets?: number[];
+  };
+}
+```
+
+### Event Interface
+
+[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments
+
+```ts
+interface Event {
+  /** The name of the event */
+  name: string;
+  /** Optional, user-friendly description for the event */
+  desc?: string;
+  /** The arguments of the event, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+  }>;
+}
+```
+
+### Type Interfaces
+
+The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.
+
+```ts
+/** An ABI-encoded type */
+type ABIType = string;
+
+/** The name of a defined struct */
+type StructName = string;
+
+/** Raw byteslice without the length prefixed that is specified in ARC-4 */
+type AVMBytes = "bytes";
+
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [fieldName: string]: ABIType | StructFields;
```

Comment:
It occurs to me that this format requires property order to be maintained to be able to map them to the underlying tuple type. Are we comfortable that will always be the case across all languages?

https://www.rfc-editor.org/rfc/rfc7159 says (emphasis mine):

>  An object is an **unordered** collection of zero or more name/value pairs, where a name is a string and a value is a string, number, boolean, null, object, or array.

Reply:
I think this a choice for the HLL. The struct have to have a specific order for ABI encoding, so the HLL can either implement an ordered struct or not expose ways of iterating keys/values (like in TEALScript)

Reply:
But this spec determines the format of the json file? If the json file is an object and the programming language parsing it doesn’t implemented ordered iteration of object keys you are hosed?

Reply:
Ah right ok I see what you are saying. So I think it should be `StructFields = { fieldName: string, type: ABIType | StructFields }[]` to preserve order as an array, right?

Reply:
Spot on I think